Build/OpenWrt.pm

```diff
@@ -0,0 +1,150 @@
+################################################################
+#
+# Copyright (c) 2017 SUSE Linux Products GmbH
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 or 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program (see the file COPYING); if not, write to the
+# Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+#
+################################################################
+
+package Build::OpenWrt;
+
+use strict;
+use warnings;
+use Build::Deb;
+use Build::Rpm;
+
+my $yamlxs = eval { require YAML::XS; $YAML::XS::LoadBlessed = 0; return 1 };
+my $yamlpp = eval { require YAML::PP; return YAML::PP->new };
+
+sub _have_yaml_parser {
+  return $yamlpp || $yamlxs ? 1 : undef;
+}
+
+sub _load_yaml {
+  my ($yaml) = @_;
+  my $data;
+  if ($yamlpp) {
+    $data = eval { $yamlpp->load_string($yaml) };
+    return $data;
+  }
+  if ($yamlxs) {
+    eval { $data = YAML::XS::Load($yaml) };
+    return $data;
+  }
+  die "Neither YAML::PP nor YAML::XS available\n";
+}
+
+sub _load_yaml_file {
+  my ($fn) = @_;
+  my $data;
+  if ($yamlpp) {
+    $data = eval { $yamlpp->load_file($fn) };
+    return $data;
+  }
+  if ($yamlxs) {
+    eval { $data = YAML::XS::LoadFile($fn) };
+    return $data;
+  }
+  die "Neither YAML::PP nor YAML::XS available\n";
+}
+
+sub _read_manifest {
+  my ($fn) = @_;
+  my $data;
+  if ($fn =~ m/\.ya?ml\z/) {
+    $data = _load_yaml_file($fn);
+    return { error => "Failed to parse YAML file '$fn'" } unless defined $data;
+  } elsif ($fn =~ m/\.json\z/) {
+    # We don't have JSON::PP, but YAML is a superset of JSON anyway
+    $data = _load_yaml_file($fn);
+    return { error => "Failed to parse JSON file '$fn'" } unless defined $data;
+  } elsif (ref($fn) eq 'SCALAR') {
+    $data = _load_yaml($$fn);		# used in the unit test
+    return { error => "Failed to parse '$fn'" } unless defined $data;
+  } else {
+    $data = _load_yaml_file($fn);
+    return { error => "Failed to parse file '$fn'" } unless defined $data;
+  }
+  return $data;
+}
```

Comment:
I'm pretty certain you don't need this part, or at least not all of it. You should be fine with a json parser.

Reply:
That makes sense, I wasn't sure whether a JSON parser was present inside the build system. I saw https://github.com/openSUSE/obs-build/blob/master/Build/SimpleJSON.pm but wasn't sure if that would work, and couldn't verify whether https://metacpan.org/dist/JSON was preinstalled.

Reply:
The nice part about this method is that it keeps yaml parsing available if someone (or something, e.g. a buildsystem) prefers YAML.